docs/reference/search/rank-eval.asciidoc

```diff
@@ -0,0 +1,297 @@
+
+=== Template based ranking evaluation
+
+As an alternative to having to provide a single query per test request, it is possible to specify query templates in the evaluation request and later refer to them. Queries with similar structure that only differ in their parameters don't have to be repeated all the time in the `requests` section this way. In typical search systems where user inputs usually get filled into a small set of query templates, this helps making the evaluation request more succinct.
+
+[source,js]
+--------------------------------
+GET /my_index/_rank_eval
+{ 
+   [...]
+  "templates": [
+     {
+        "id": "match_one_field_query",  <1>
+        "template": { <2>
+            "inline": { 
+                "query": { 
+                  "match": { "{{field}}": { "query": "{{query_string}}" }}
+                }
+            }
+        }
+     }
+  ], 
+  "requests": [
+      {
+         "id": "amsterdam_query"
+         "ratings": [ ... ],
+         "template_id": "match_one_field_query", <3>
+         "params": { <4>
+            "query_string": "amsterdam",
+            "field": "text"
+          }
+     },
+    [...]
+}
+--------------------------------
+// NOTCONSOLE
+
+<1> the template id 
+<2> the template definition to use
+<3> a reference to a previously defined temlate
+<4> the parameters to use to fill the template
+
+=== Available evaluation metrics
+
+The `metric` section determines which of the available evaluation metrics is going to be used.
+Currently, the following metrics are supported:
+
+==== Precision at K (P@k)
+
+This metric measures the number of relevant results in the top k search results. Its a form of the well known https://en.wikipedia.org/wiki/Information_retrieval#Precision[Precision] metric that only looks at the top k documents. It is the fraction of relevant documents in those first k
+search. A precision at 10 (P@10) value of 0.6 then means six out of the 10 top hits are relevant with respect to the users information need.
+
+P@k works well as a simple evaluation metric that has the benefit of being easy to understand and explain.
+Documents in the collection need to be rated either as relevant or irrelevant with respect to the current query. 
+P@k does not take into account where in the top k results the relevant documents occur, so a ranking of ten results that 
+contains one relevant result in position 10 is equally good as a ranking of ten results that contains one relevant result in position 1.
+
+[source,js]
+--------------------------------
+GET /twitter/_rank_eval
+{
+    "requests": [
+    {
+        "id": "JFK query", 
+        "request": { "query": { "match_all": {}}},
+        "ratings": []  
+    }],
+    "metric": {
+      "precision": {
+        "relevant_rating_threshold": 1,
+        "ignore_unlabeled": false
+      }
+   }
+}
+--------------------------------
+// CONSOLE
+// TEST[setup:twitter]
+
+The `precision` metric takes the following optional parameters
+
```

Comment:
Is there a parameter for `k` in the `P@k`, or `k` is always equal to 10?

Reply:
We had an expiclit parameter at some point but removed it since the size of the response can be controlled easily using the `size` parameter in each search request body (the `request` section accepts a full search request body). Thinking about it now, it might be a good idea to change it back though, because in the currently one would have to add the "size" to every query in the test suite (and could accidentally use different sizes), so maybe we should add this option back. 
Thanks for the input, will think about this again.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/27569 as an example how adding `k` might look like. We can pull it in now or use it for further discussion.